core/market-decentralized/src/market.rs

```diff
@@ -0,0 +1,122 @@
+use lazy_static::lazy_static;
+use std::env::current_dir;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use thiserror::Error;
+
+use crate::matcher::{Matcher, MatcherInitError};
+use crate::negotiation::NegotiationInitError;
+use crate::negotiation::{ProviderNegotiationEngine, RequestorNegotiationEngine};
+use crate::protocol::{DiscoveryBuilder, DiscoveryGSB};
+
+use ya_core_model::market::BUS_ID;
+use ya_persistence::executor::DbExecutor;
+use ya_service_api_interfaces::{Provider, Service};
+
+#[derive(Error, Debug)]
+pub enum MarketError {}
+
+#[derive(Error, Debug)]
+pub enum MarketInitError {
+    #[error("Failed to initialize Offers matcher. Error: {}.", .0)]
+    Matcher(#[from] MatcherInitError),
+    #[error("Failed to initialize negotiation engine. Error: {}.", .0)]
+    Negotiation(#[from] NegotiationInitError),
+}
+
+/// Structure connecting all market objects.
+pub struct Market {
+    matcher: Arc<Matcher>,
+    provider_negotiation_engine: Arc<ProviderNegotiationEngine>,
+    requestor_negotiation_engine: Arc<RequestorNegotiationEngine>,
+}
+
+impl Market {
+    pub fn new(db: &DbExecutor, data_dir: &Path) -> Result<Self, MarketInitError> {
+        // TODO: Set Matcher independent parameters here or remove this todo.
+        let builder = DiscoveryBuilder::new();
+
+        let (matcher, listeners) = Matcher::new::<DiscoveryGSB>(builder, data_dir)?;
+        let provider_engine = ProviderNegotiationEngine::new(db.clone())?;
+        let requestor_engine =
+            RequestorNegotiationEngine::new(db.clone(), listeners.proposal_receiver)?;
+
+        Ok(Market {
+            matcher: Arc::new(matcher),
+            provider_negotiation_engine: provider_engine,
+            requestor_negotiation_engine: requestor_engine,
+        })
+    }
+
+    pub async fn bind_gsb(&self, prefix: String) -> Result<(), MarketInitError> {
+        self.matcher.bind_gsb(prefix.clone()).await?;
+        self.provider_negotiation_engine
+            .bind_gsb(prefix.clone())
+            .await?;
+        self.requestor_negotiation_engine.bind_gsb(prefix).await?;
+        Ok(())
+    }
+
+    pub async fn gsb<Context: Provider<Self, DbExecutor> + Provider<Self, PathBuf>>(
+        ctx: &Context,
+    ) -> anyhow::Result<()> {
+        let db = ctx.component();
+        let data_dir: PathBuf = ctx.component();
+        let market = MARKET.get_or_init_market(&db, &data_dir)?;
+        Ok(market.bind_gsb(BUS_ID.to_string()).await?)
+    }
+
+    pub fn rest(db: &DbExecutor) -> actix_web::Scope {
+        // TODO: We should run gsb function first, because it gets datadir.
```

Comment:
It might be good to made `rest` and `gsb` consistent in sens they both have Context on input. But definitely not now